syntax = "proto3";

package fleet;

option go_package = "github.com/greenlane/proto/fleet";

// Service definition for bidirectional streaming between EVs and the ingestion service
service FleetService {
  // Bidirectional stream: EVs send telemetry, server sends booking confirmations
  rpc StreamTelemetry(stream CarStatus) returns (stream BookingResponse);
}

// CarStatus represents real-time telemetry from an EV
message CarStatus {
  string car_id = 1;              // Unique car identifier
  double latitude = 2;            // Current latitude
  double longitude = 3;           // Current longitude
  double battery_level = 4;       // Battery percentage (0-100)
  double velocity = 5;            // Speed in km/h
  int64 timestamp = 6;            // Unix timestamp (milliseconds)
}

// BookingRequest is sent when a car wants to reserve a charging slot
message BookingRequest {
  string car_id = 1;              // Car making the request
  string station_id = 2;          // Target charging station
  int64 expected_arrival = 3;     // Expected arrival time (Unix timestamp, milliseconds)
  double battery_level = 4;       // Current battery level
}

// BookingResponse is sent back to confirm or deny a booking
message BookingResponse {
  string booking_id = 1;          // Unique booking identifier
  string car_id = 2;              // Car this response is for
  string station_id = 3;          // Assigned station
  BookingStatus status = 4;       // Status of the booking
  string message = 5;             // Human-readable message
  int64 timestamp = 6;            // Response timestamp
  double price_rate = 7;          // Price per kWh at booking time
}

// BookingStatus enum
enum BookingStatus {
  BOOKING_UNKNOWN = 0;
  BOOKING_CONFIRMED = 1;
  BOOKING_FAILED = 2;
  BOOKING_EXPIRED = 3;
}

// Event messages for Kafka/Redpanda
message TelemetryEvent {
  string car_id = 1;
  double latitude = 2;
  double longitude = 3;
  double battery_level = 4;
  double velocity = 5;
  int64 timestamp = 6;
  string event_type = 7;          // "telemetry", "booking_confirmed", "booking_failed"
}

message BookingEvent {
  string booking_id = 1;
  string car_id = 2;
  string station_id = 3;
  BookingStatus status = 4;
  int64 timestamp = 5;
  double price_rate = 6;
  string event_type = 7;          // "booking_confirmed", "booking_failed", "charging_started"
}
